### ISSUE TITLE
Prisma can't run in WebContainer

### ISSUE THREAD

**Issue by pegak** [prisma/prisma#17710]
### Bug description

https://github.com/stackblitz/webcontainer-core/issues/954

### How to reproduce

1. Go to Stackblitz Codeflow
2. Use Prisma in deps
3. Run pnpm

### Prisma information

```prisma
// Add your schema.prisma
```

```ts
// Add your code using Prisma Client
```

### Environment & setup

- OS: 
- Database: 
- Node.js version: 

### Prisma Version

```

```

**Comment by janpio** [comment]
Related and some background discussion: https://github.com/prisma/prisma/issues/8626

We can keep this issue though to track the general problem as a bug.

**Comment by pegak** [comment]
Thank you for such a fast reply. I will read the thread. :)

**Comment by janpio** [comment]
Simple and straightforward reproduction: https://stackblitz.com/~/github.com/janpio/stackblitz-codeflow-prisma

Some background info from Stackblitz on their side: https://github.com/stackblitz/webcontainer-core/issues/954#issuecomment-1417709580

> Regarding "socket hang up", that's due to CORS because [binaries.prisma.sh](https://binaries.prisma.sh/) is blocking requests from *.staticblitz.com. If you can enable CORS for that domain the socket shouldn't hang up anymore.

Unfortunately that will still not let it execute the engine after downloading of course.

**Comment by janpio** [comment]
This Chrome extension is a simple way to emulate CORS being enabled (careful, only keep active while testing): https://chrome.google.com/webstore/detail/cors-unblock/lfhmikememgdcahcdlaciloancbhjino

With it enabled, Stackblit can download the engines, but still not execute them of course. Here is the output of `prisma -v` in that situation:

```
~/janpio/stackblitz-codeflow-prisma main
❯ npx prisma -v
prisma:warn Prisma failed to detect the libssl/openssl version to use, and may not work as expected. Defaulting to "openssl-1.1.x".
Please manually install OpenSSL and try installing Prisma again.
prisma:warn Prisma doesn't know which engines to download for the Linux distro "undefined". Falling back to Prisma engines built "debian".
Please report your experience by creating an issue at https://github.com/prisma/prisma/issues so we can add your distro to the list of known supported distros.
prisma:warn Prisma failed to detect the libssl/openssl version to use, and may not work as expected. Defaulting to "openssl-1.1.x".
Please manually install OpenSSL and try installing Prisma again.
prisma:warn Prisma doesn't know which engines to download for the Linux distro "undefined". Falling back to Prisma engines built "debian".
Please report your experience by creating an issue at https://github.com/prisma/prisma/issues so we can add your distro to the list of known supported distros.
Error: Unable to require(`/home/janpio/stackblitz-codeflow-prisma/node_modules/@prisma/engines/libquery_engine-debian-openssl-1.1.x.so.node`).
The Prisma engines do not seem to be compatible with your system. Please refer to the documentation about Prisma's system requirements: https://pris.ly/d/system-requirements

Details: Cannot load native addon because loading addons is disabled: /home/janpio/stackblitz-codeflow-prisma/node_modules/@prisma/engines/libquery_engine-debian-openssl-1.1.x.so.node
    at loadLibrary (file:///home/janpio/stackblitz-codeflow-prisma/node_modules/prisma/build/index.js:93541:11)
    at getEngineVersion (file:///home/janpio/stackblitz-codeflow-prisma/node_modules/prisma/build/index.js:93555:16)
Error: Command failed with exit code 2: /home/janpio/stackblitz-codeflow-prisma/node_modules/@prisma/engines/migration-engine-debian-openssl-1.1.x --version
/home/janpio/stackblitz-codeflow-prisma/node_modules/@prisma/engines/migration-engine-debian-openssl-1.1.x: line 2: a command can only contain words and redirects
    at makeError (file:///home/janpio/stackblitz-codeflow-prisma/node_modules/prisma/build/index.js:3640:18)
    at handlePromise (file:///home/janpio/stackblitz-codeflow-prisma/node_modules/prisma/build/index.js:4408:33)
    at async getEngineVersion (file:///home/janpio/stackblitz-codeflow-prisma/node_modules/prisma/build/index.js:93558:20) {
  shortMessage: 'Command failed with exit code 2: /home/janpio/stackblitz-codeflow-prisma/node_modules/@prisma/engines/migration-engine-debian-openssl-1.1.x --version',
  command: '/home/janpio/stackblitz-codeflow-prisma/node_modules/@prisma/engines/migration-engine-debian-openssl-1.1.x --version',
  escapedCommand: '"/home/janpio/stackblitz-codeflow-prisma/node_modules/@prisma/engines/migration-engine-debian-openssl-1.1.x" --version',
  exitCode: 2,
  signal: undefined,
  signalDescription: undefined,
  stdout: '',
  stderr: '/home/janpio/stackblitz-codeflow-prisma/node_modules/@prisma/engines/migration-engine-debian-openssl-1.1.x: line 2: a command can only contain words and redirects',
  failed: true,
  timedOut: false,
  isCanceled: false,
  killed: false
}
prisma                  : 4.10.1
@prisma/client          : Not found
Current platform        : debian-openssl-1.1.x
Query Engine (Node-API) : E_CANNOT_RESOLVE_VERSION (at node_modules/@prisma/engines/libquery_engine-debian-openssl-1.1.x.so.node)
Migration Engine        : E_CANNOT_RESOLVE_VERSION (at node_modules/@prisma/engines/migration-engine-debian-openssl-1.1.x)
Format Wasm             : @prisma/prisma-fmt-wasm 4.10.1-1.80b351cc7c06d352abe81be19b8a89e9c6b7c110
Default Engines Hash    : aead147aa326ccb985dcfed5b065b4fdabd44b19
Studio                  : 0.481.0
```

The relevant error messages for Query Engine node-API libary:
```
Error: Unable to require(`/home/janpio/stackblitz-codeflow-prisma/node_modules/@prisma/engines/libquery_engine-debian-openssl-1.1.x.so.node`).
The Prisma engines do not seem to be compatible with your system. Please refer to the documentation about Prisma's system requirements: https://pris.ly/d/system-requirements

Details: Cannot load native addon because loading addons is disabled: /home/janpio/stackblitz-codeflow-prisma/node_modules/@prisma/engines/libquery_engine-debian-openssl-1.1.x.so.node
    at loadLibrary (file:///home/janpio/stackblitz-codeflow-prisma/node_modules/prisma/build/index.js:93541:11)
    at getEngineVersion (file:///home/janpio/stackblitz-codeflow-prisma/node_modules/prisma/build/index.js:93555:16)
Error: Command failed with exit code 2: /home/janpio/stackblitz-codeflow-prisma/node_modules/@prisma/engines/migration-engine-debian-openssl-1.1.x --version
```

and Migration Engine binary:

```
/home/janpio/stackblitz-codeflow-prisma/node_modules/@prisma/engines/migration-engine-debian-openssl-1.1.x: line 2: a command can only contain words and redirects
    at makeError (file:///home/janpio/stackblitz-codeflow-prisma/node_modules/prisma/build/index.js:3640:18)
    at handlePromise (file:///home/janpio/stackblitz-codeflow-prisma/node_modules/prisma/build/index.js:4408:33)
    at async getEngineVersion (file:///home/janpio/stackblitz-codeflow-prisma/node_modules/prisma/build/index.js:93558:20) {
  shortMessage: 'Command failed with exit code 2: /home/janpio/stackblitz-codeflow-prisma/node_modules/@prisma/engines/migration-engine-debian-openssl-1.1.x --version',
  command: '/home/janpio/stackblitz-codeflow-prisma/node_modules/@prisma/engines/migration-engine-debian-openssl-1.1.x --version',
  escapedCommand: '"/home/janpio/stackblitz-codeflow-prisma/node_modules/@prisma/engines/migration-engine-debian-openssl-1.1.x" --version',
  exitCode: 2,
  signal: undefined,
  signalDescription: undefined,
  stdout: '',
  stderr: '/home/janpio/stackblitz-codeflow-prisma/node_modules/@prisma/engines/migration-engine-debian-openssl-1.1.x: line 2: a command can only contain words and redirects',
  failed: true,
  timedOut: false,
  isCanceled: false,
  killed: false
}
```

And for completeness, also the Query Engine binary (via `PRISMA_CLI_QUERY_ENGINE_TYPE=binary`):

```
Error: Command failed with exit code 2: /home/janpio/stackblitz-codeflow-prisma/node_modules/@prisma/engines/query-engine-debian-openssl-1.1.x --version
/home/janpio/stackblitz-codeflow-prisma/node_modules/@prisma/engines/query-engine-debian-openssl-1.1.x: line 4: a command can only contain words and redirects
    at makeError (file:///home/janpio/stackblitz-codeflow-prisma/node_modules/prisma/build/index.js:3640:18)
    at handlePromise (file:///home/janpio/stackblitz-codeflow-prisma/node_modules/prisma/build/index.js:4408:33)
    at async getEngineVersion (file:///home/janpio/stackblitz-codeflow-prisma/node_modules/prisma/build/index.js:93558:20) {
  shortMessage: 'Command failed with exit code 2: /home/janpio/stackblitz-codeflow-prisma/node_modules/@prisma/engines/query-engine-debian-openssl-1.1.x --version',
  command: '/home/janpio/stackblitz-codeflow-prisma/node_modules/@prisma/engines/query-engine-debian-openssl-1.1.x --version',
  escapedCommand: '"/home/janpio/stackblitz-codeflow-prisma/node_modules/@prisma/engines/query-engine-debian-openssl-1.1.x" --version',
  exitCode: 2,
  signal: undefined,
  signalDescription: undefined,
  stdout: '',
  stderr: '/home/janpio/stackblitz-codeflow-prisma/node_modules/@prisma/engines/query-engine-debian-openssl-1.1.x: line 4: a command can only contain words and redirects',
  failed: true,
  timedOut: false,
  isCanceled: false,
  killed: false
}
```

**Comment by janpio** [comment]
(Technically we would now be in a position to `prisma generate --data-proxy` and then use that in an app to send queries via Data Proxy. Currently the Data Proxy does not set any CORS headers and does not respond to preflight requests though - starting a discussion about that internally.)

**Comment by GauBen** [comment]
Hey there! I'm also interested in running prisma+sqlite in a webcontainer. <https://binaries.prisma.sh> looks like a regular aws s3, is there a reason not to [set permissive AllowedOrigins?](https://docs.aws.amazon.com/AmazonS3/latest/userguide/ManageCorsUsing.html)

**Comment by GauBen** [comment]
I'm building an in-browser demo of graphql+prisma aimed at a (relatively) large audience of beginners, so I'd rather avoid telling them to install a browser extension designed to reduce the security of their browser, but I will go this route for now

> Yes, it is not needed for Prisma's normal use case

We might consider installing a WASM build of prisma a normal use case, as WASM is now officially supported

**Comment by cassepipe** [comment]
I am trying to a fullstack [project](https://stackblitz.com/edit/stackblitz-webcontainer-api-starter-d8ux9k?file=prisma-generated%2Fpackage.json,package.json) on Stackblitz too and if you go in the prisma-generated directory and run
`pnpm prisma generate`, you will see that command never gives you your prompt back.

**Comment by cassepipe** [comment]
Anyone managed to get `prisma` to install, generate a client and push to a db successfully from StackBlitz yet ?

**Comment by janpio** [comment]
Normal Prisma is not possible in Webcontainers because of this limitation: https://webcontainers.io/guides/troubleshooting#cannot-load-native-addon-error

There might be Wasm-based build of Prisma for specific databases soon, that might circumvent this problem.

**Comment by giovanniincammicia** [comment]
Any update on this? I tried `prisma generate --data-proxy` but it still doesn't work for me. Any workaround?

**Comment by yasinkocak** [comment]
Facing same issue, is there workaround, chrome extension didn't work

**Comment by BLew10** [comment]
Also facing same issue still a year a later.

**Comment by Shabaz14** [comment]
same here facing this issue , any solution

**Comment by copostic** [comment]
same here... Still no solution ?
